### PR TITLE
Markdown Rendering in the What's New Panel

### PR DESCRIPTION
**Pull request by kateliu20**
This is a PR for markdown file (.md) rendering and formatting for the "What's New at Slack!" side panel in the Skate plugin. This change went through a few iterations because I was having trouble using intelliJ's markdown plugin in Skate, so I had done custom CSS formatting before. However, now I was able to use MarkdownJCEFHtmlPanel to properly format the md files. Here is what it looks like as it depends on the formatting for the markdown plugin with intelliJ:

Screenshot 2023-07-06 at 5 31 07 PM

Screenshot 2023-07-06 at 6 02 59 PM

### CODE REVIEW

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt [hunk 1]
```diff
@@ -18,20 +18,25 @@ package com.slack.sgp.intellij
...
+ * New UI of the Skate Plugin
+ */
 class SkateProjectServiceImpl(private val project: Project) : SkateProjectService {
 
   override fun showWhatsNewWindow() {
     // TODO
-    //  Make the file configurable?
     //  Only show when changed
     //  Only show latest changes
     val settings = project.service<SkatePluginSettings>()
```
Comment (ZacSweers) [inline skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt:42]: Can do this in a separate PR but let's make sure we connect these pieces to only show changes since the last update

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt [hunk 2]
```diff
@@ -46,7 +51,11 @@ class SkateProjectServiceImpl(private val project: Project) : SkateProjectServic
           stripeTitle = Supplier { "What's New in Slack!" }
           anchor = ToolWindowAnchor.RIGHT
         }
-      WhatsNewPanelFactory().createToolWindowContent(toolWindow, changeLogString)
+
+      val parentDisposable = Disposer.newDisposable() // Creating a new Disposable
```
Comment (ZacSweers) [inline skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt:55]: This comment is a bit redundant. Instead, let's say what it's for.

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    // Control Panel that takes in the current project, markdown string, and a Disposable.
+    // The Disposable is necessary to prevent a substantial memory leak while working with
+    // MarkdownJCEFHtmlPanel
+    private fun createControlsPanel(
+      project: Project,
+      @Language("Markdown") markdownFileString: String,
+      parentDisposable: Disposable
+    ): JComponent {
+      println(markdownFileString)
```
Comment (ZacSweers) [inline skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt:75]: nit: let's remove this

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      val file = LightVirtualFile("changelog.md", markdownFileString)
+      val panel = MarkdownJCEFHtmlPanel(project, file)
+      Disposer.register(parentDisposable, panel)
+      val html = runReadAction {
```
Comment (ZacSweers) [inline skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt:79]: for my own knowledge, what's `runReadAction` do?

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    val contentPanel: JPanel =
+      JPanel().apply {
+        layout = BorderLayout(0, 20)
+        add(createControlsPanel(project, markdownFileString, parentDisposable), BorderLayout.CENTER)
+      }
+
+    // Control Panel that takes in the current project, markdown string, and a Disposable.
+    // The Disposable is necessary to prevent a substantial memory leak while working with
+    // MarkdownJCEFHtmlPanel
+    private fun createControlsPanel(
```
Comment (ZacSweers) [inline skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt:70]: Wouldn't this be the "whats new" panel vs control panel?

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    // Actual panel box for What's New at Slack
+    val contentPanel: JPanel =
+      JPanel().apply {
+        layout = BorderLayout(0, 20)
+        add(createControlsPanel(project, markdownFileString, parentDisposable), BorderLayout.CENTER)
+      }
+
+    // Control Panel that takes in the current project, markdown string, and a Disposable.
+    // The Disposable is necessary to prevent a substantial memory leak while working with
```
Comment (ZacSweers) [inline skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt:68]: We don't do anything with the parent disposable above though, could you elaborate on the leak it's preventing and how?

File: skate-plugin/build.gradle.kts [hunk 7]
```diff
@@ -38,3 +38,7 @@ dependencies {
   testImplementation(libs.junit)
   testImplementation(libs.truth)
 }
+java {
+    sourceCompatibility = JavaVersion.VERSION_19
+    targetCompatibility = JavaVersion.VERSION_19
+}
```
[the remark this answers is not in the document]
Comment (kateliu20) [reply skate-plugin/build.gradle.kts]: Yeah, I keep on seeing this in every PR I make. I don't know why it keeps popping up? I think I deleted it though.

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt (hunk 1 above)
Comment (kateliu20) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt:42]: Yes for sure, should I delete these todo comments for now?

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt (hunk 6 above)
Comment (kateliu20) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt:68]: I had actually gotten that information from here: https://intellij-support.jetbrains.com/hc/en-us/community/posts/10856308259090-Markdown-Preview-Swing-Component. This is also where I got how to implement MarkdownJCEFHtmlPanel.

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt (hunk 4 above)
Comment (kateliu20) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt:79]: I believe that `runReadAction` is so we can read the markdown content and convert it to html. Here is some documentation for it too: https://plugins.jetbrains.com/docs/intellij/general-threading-rules.html#read-access, and it's used in: https://intellij-support.jetbrains.com/hc/en-us/community/posts/10856308259090-Markdown-Preview-Swing-Component.

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt (hunk 5 above)
Comment (kateliu20) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt]: Changed it in here: ef81fb1e30983db32d66a35bdaf08df2bd0a5e5a

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt (hunk 3 above)
Comment (kateliu20) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt]: Removed it: ef81fb1e30983db32d66a35bdaf08df2bd0a5e5a

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt (hunk 2 above)
Comment (kateliu20) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt]: Sure, I added the comment there: ef81fb1e30983db32d66a35bdaf08df2bd0a5e5a

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt (hunk 1 above)
Comment (ZacSweers) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/SkateService.kt:42]: I'd say leave 'em until we actually connect the parser to the previously seen state

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt (hunk 6 above)
Comment (ZacSweers) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt]: So this only prevents a leak if we dispose the panel somewhere, but I don't see that happening in this code. It looks like we only create the disposable. A disposable is a common concept used for cases where you have some process or resource and you want to be able to "dispose" of it later. We use it a lot in RxJava too. It's only doing its job if we call `dispose()` somewhere though, so I think we still have more to do here

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt (hunk 4 above)
Comment (ZacSweers) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt:76]: Got it, so looking at that first one it's more that it guarantees safe access if we're not on the main thread?

File: skate-plugin/build.gradle.kts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ZacSweers) [reply skate-plugin/build.gradle.kts]: It's something with your IDE settings, you should be using JDK 20 but your IDE keeps trying to change this to JDK 19

File: skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt (hunk 6 above)
Comment (kateliu20) [reply skate-plugin/src/main/kotlin/com/slack/sgp/intellij/ui/WhatsNewPanelFactory.kt]: Should I be disposing it in `SkateService` after the tool window is closed maybe?

**Review by ZacSweers — APPROVED** [review]
